gix-archive/src/archive.rs

```diff
@@ -0,0 +1,44 @@
+use std::io::{Read, Write};
+use std::path::PathBuf;
+use std::slice::Iter;
+
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+pub enum Error {
+    #[error("Failed to archive")]
+    FailedToArchive,
+}
+
+/// File to be added to the archive.
+/// It is processed (contents after applying .gitattributes/filtering)
+/// TODO Do we need to care about empty dirs and links?
+pub struct File<R: Read> {
+    pub path: PathBuf,
+    pub data: R,
+}
+
+/// Available formats
+#[derive(PartialEq)]
+pub enum Format {
+    Tar,
+    Zip,
+}
+
+/// Archive options, such as format, and later prefix, work-tree, etc
+pub struct Options {
+    pub format: Format,
+}
+
+/// Iterates over `_entries` (ready to be added to an archive files) and writes to `_destination` according to archive `opts`
+pub fn archive<R: Read, W: Write>(
+    _entries: Iter<'_, File<R>>,
```

Comment:
Doing it like this means the data cannot be streamed, and everything has to be in memory in advance, thus doesn't scale.

Reply:
I thought that since iter is lazy, only a single file has to be in mem, or not?

Reply:
It's hard-coded to be a slice iterator, which implies all of it is in memory in advance. What's seems to be intended is `impl IntoIterator<Item = File<R>>`. I see issues happening when trying to fill in `R` with an actual type and recommend using a `Vec<u8>` instead - I consider huge file support optional as `gitoxide` wouldn't be good with huge files as well. Ideally there is a way to reuse these data `Vec`s as well, but I think once things became clearer it will get there naturally.

Reply:
ok, changed to `impl IntoIterator<File>`(File has Vec<u8>, not Read)

Reply:
> I see issues happening when trying to fill in `R` with an actual type

What issues are you seeing?